doc/neps/nep-0031-uarray.rst

```diff
@@ -0,0 +1,387 @@
+``uarray``, that is the purpose of the ``uarray`` documentation.* `[1]`_
+*However, the NumPy community will have input into the design of
+``uarray``, via the issue tracker.*
+
+``uarray`` Primer
+^^^^^^^^^^^^^^^^^
+
+Defining backends
+~~~~~~~~~~~~~~~~~
+
+``uarray`` consists of two main protocols: ``__ua_convert__`` and
+``__ua_function__``, called in that order, along with ``__ua_domain__``, which
+is a string defining the domain of the backend. If any of the protocols return
+``NotImplemented``, we fall back to the next backend.
+
+``__ua_convert__`` is for conversion and coercion. It has the signature
+``(dispatchables, coerce)``, where ``dispatchables`` is an iterable of
+``ua.Dispatchable`` objects and ``coerce`` is a boolean indicating whether or
+not to force the conversion. ``ua.Dispatchable`` is a simple class consisting
+of three simple values: ``type``, ``value``, and ``coercible``.
+``__ua_convert__`` returns an iterable of the converted values, or
+``NotImplemented`` in the case of failure. Returning ``NotImplemented``
+here will cause ``uarray`` to move to the next available backend.
+
+``__ua_function__`` has the signature ``(func, args, kwargs)`` and defines
+the actual implementation of the function. It recieves the function and its
+arguments. Returning ``NotImplemented`` will cause a move to the default
+implementation of the function if one exists, and failing that, the next
+backend.
+
+If all backends are exhausted, a ``ua.BackendNotImplementedError`` is raised.
+
+Backends can be registered for permanent use if required.
+
+Defining overridable methods
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To define an overridable method, one needs a few things:
+
+1. A dispatcher that returns an iterable of ``ua.Dispatchable`` objects.
+2. A reverse dispatcher that replaces dispatchable values with the supplied
+   ones.
+3. A domain.
+4. Optionally, a default implementation, which can be provided in terms of
+   other overridable methods.
+
+As an example, consider the following::
+
+    import uarray as ua
+
+    def full_argreplacer(args, kwargs, dispatchables):
+        def full(shape, fill_value, dtype=None, order='C'):
+            return (shape, fill_value), dict(
+                dtype=dispatchables[0],
+                order=order
+            )
+
+        return full(*args, **kwargs)
+
+    @ua.create_multimethod(full_argreplacer, domain="numpy")
+    def full(shape, fill_value, dtype=None, order='C'):
+        return (ua.Dispatchable(dtype, np.dtype),)
+
+A large set of examples can be found in the ``unumpy`` repository, `[8]`_.
+This simple act of overriding callables allows us to override:
+
+* Methods
+* Properties, via ``fget`` and ``fset``
+* Entire objects, via ``__get__``.
+
+Using overrides
+~~~~~~~~~~~~~~~
+
+The way we propose the overrides will be used by end users is::
+
+    import numpy.overridable as np
+    with np.set_backend(backend):
+        x = np.asarray(my_array, dtype=dtype)
```

Comment:
What if the library requires, or at least accepts, more parameters than numpy? e.g. 

- dask -> chunks
- sparse -> fill_value
- cupy -> device? (atm it is set with a context manager or global variable)

Reply:
If that's the case, one can simply use methods from other libraries. If we accept keyword arguments that are arbitrary and drop them later, it may lead to incorrect results. One resolution is hint keyword arguments: Quansight-Labs/uarray#164